providers/base/bin/prime_offload_tester.py

```diff
@@ -0,0 +1,357 @@
+#!/usr/bin/env python3
+# This file is part of Checkbox.
+#
+# Copyright 2023 Canonical Ltd.
+# Written by:
+#   Hanhsuan Lee <hanhsuan.lee@canonical.com>
+#
+# Checkbox is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3,
+# as published by the Free Software Foundation.
+#
+# Checkbox is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Checkbox.  If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+import threading
+import subprocess
+import time
+import re
+import json
+import argparse
+import logging
+import os
+
+
+class PrimeOffloader:
+    """
+    A class used to execute process to specific GPU.
+    Have to run this as root.
+
+    :attr logger: console logger
+    :type logger: obj
+
+    :attr check_result:
+        store the result of checking offloading is ok or not.
+    :type check_result: int
+    """
+
+    logger = logging.getLogger()
+    check_result = False
+
+    def find_card_id(self, pci_name):
+        """
+        use pci name to find card id under /sys/kernel/debug/dri
+
+        :param pci_name: pci device name in NNNN:NN:NN.N format
+        :type card_name: str
+
+        : returns : card id
+        : rtype : str
+        """
+        if not re.match("[0-9]{4}:[0-9]{2}:[0-9]{2}.[0-9]", pci_name):
+            raise RuntimeError("pci name format error")
+
+        try:
+            cmd = ["grep",
+                   "-lr",
+                   "--include=name",
+                   pci_name,
+                   "/sys/kernel/debug/dri"]
+
+            card_path = subprocess.check_output(cmd,
+                                                universal_newlines=True)
+            return card_path.split('/')[5]
+        except (IndexError, AttributeError) as e:
+            raise RuntimeError("return value format error {}".format(e))
```

Comment:
the inner raise creates an exception raising context while already inside a raising context. What it means the program crashes with the double traceback, like so:
```
Traceback (most recent call last):
  File "/home/kissiel/tmp/exc.py", line 7, in <module>
    foo()
  File "/home/kissiel/tmp/exc.py", line 3, in foo
    return a[5]
           ~^^^
IndexError: list index out of range

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kissiel/tmp/exc.py", line 9, in <module>
    raise RuntimeError("error msg {}".format(e))
RuntimeError: error msg list index out of range
```
The code that caused that:
```
def foo():
    a = [1]
    return a[5]


try:
    foo()
except IndexError as e:
    raise SystemExit("error msg {}".format(e)) 
```

If you're fine with crashing the program here, the original exception is fine, the double exception IMHO doesn't introduce anything of value and just obfuscates the code.

Of course gracefully exiting with SystemExit would be a much better solution.